src/critical_section.c

```diff
@@ -0,0 +1,54 @@
+// Copyright (c) Microsoft. All rights reserved.
+#include <stdint.h>
+
+#include "c_pal/sync.h"
+#include "c_pal/interlocked.h"
+#include "c_pal/log_critical_and_terminate.h"
+
+#include "c_util/critical_section.h"
+
+void enter_crit_section(volatile_atomic int32_t* access_value)
+{
+    if (access_value == NULL)
+    {
+        /*Codes_SRS_CRITICAL_SECTION_18_001: [ If access_value is NULL, enter_crit_section shall terminate the process. ]*/
+        LogCriticalAndTerminate("Invalid args: volatile_atomic int32_t* access_value=%p", access_value);
+    }
+    else
+    {
+        do
+        {
+            /*Codes_SRS_CRITICAL_SECTION_18_002: [ enter_crit_section shall call interlocked_compare_exchange to set access_value to 1 if it was previously 0. ]*/
+            int32_t current_val = interlocked_compare_exchange(access_value, 1, 0);
+            /*Codes_SRS_CRITICAL_SECTION_18_003: [ If interlocked_compare_exchange indicates that access_value was changed from 0 to 1, enter_crit_section returns. ]*/
+            if (current_val == 0)
+            {
+                break;
+            }
+            else
+            {
+                // Do Nothing wait for address
+            }
```

Comment:
I would actually put the wait_on_address in this else branch, then there is no need for the extra comment of do nothing